Core/Util.lua

```diff
@@ -141,3 +141,42 @@ end
 function Util.ParseQuality(value)
     return value and PET_QUALITIES[lower(value)] or nil
 end
+
+-- See Interface/FrameXML/SharedPetBattleTemplates.lua's SharedPetAbilityTooltip_ParseExpressionInsecure for how this works.
+local mathParserEnv = {
+    ceil = math.ceil,
+    floor = math.floor,
+    abs = math.abs,
+    min = math.min,
+    max = math.max,
+    cond = function(conditional, onTrue, onFalse) if ( conditional ) then return onTrue; else return onFalse; end end,
+    clamp = function(value, minClamp, maxClamp) return min(max(value, minClamp), maxClamp); end,
+}
+
+local mathParserSafeEnv = {}
+setmetatable(mathParserSafeEnv, { __index = mathParserEnv, __newindex = function() end, __metatable = false })
+
+local function mathParseExpression(expression)
+    forceinsecure()
+
+    local expr = loadstring("return ("..expression..")")
```

Comment:
I feel this is not the safest way to do this. I would prefer writing a manual parser. While it would take more time to do, we would have proper control over user input.

Reply:
Blizzard trusts it at least, then again, they have all input under control.

`loadstring()` doesn't execute anything yet. `setfenv()` *completely* replaces the environment. The only things available within that environment are things defined within that expression, and in the given table. So only `mathParserEnv` is available in there.

`__index` gives only those functions that are listed with no way to modify. `__newindex = function() end` prevents adding anything to that table from within. `__metatable` prevents changing the metatable using `setmetatable()`.

The internet suggests to not provide `getfenv()`, `getmetatable()` and `debug` inside the environment, which we don't. 

Tl;dr: This should be perfectly safe.

Reply:
Does it still allow arbitrary execution of LUA code that is placed in a pet battle script? I am convinced it is safe, so thanks for the explanation! 😄But it would be quite odd if we start seeing scripts that contain LUA which this line executes, even if safe.

Reply:
Should mathParseExpression check that what's returned is actually a number? That way even if people try to get up to funny business with LUA they shouldn't be able to return anything we don't expect.

Reply:
We can add that, I don’t have hard feelings. We currently only have this for mathy stuff but technically we could use it for other conditions as well where non-number return values make sense.

Just slap `tonumber()` around the last return?